app/src/main/java/bean/pwr/imskamieskiego/MapGenerator/BitmapDecoder.java

```diff
@@ -0,0 +1,37 @@
+package bean.pwr.imskamieskiego.MapGenerator;
+
+import android.content.res.Resources;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+
+public abstract class BitmapDecoder {
+
+    public static int calculateInSampleSize(
+            BitmapFactory.Options options, int reqWidth, int reqHeight) {
+        final int height = options.outHeight;
+        final int width = options.outWidth;
+        int inSampleSize = 1;
+
+        if (height > reqHeight || width > reqWidth) {
+
+            final int halfHeight = height / 2;
+            final int halfWidth = width / 2;
+            while ((halfHeight / inSampleSize) >= reqHeight
+                    && (halfWidth / inSampleSize) >= reqWidth) {
+                inSampleSize *= 2;
+            }
+        }
+
+        return inSampleSize;
+    }
+    public static Bitmap decodeSampledBitmapFromResource(Resources res, int resId,
+                                                         int reqWidth, int reqHeight) {
+
+        final BitmapFactory.Options options = new BitmapFactory.Options();
+        options.inJustDecodeBounds = true;
+        BitmapFactory.decodeResource(res, resId, options);
```

Comment:
Dlaczego to wykonuje się dwa razy? Teraz i 3 linijki dalej?

Reply:
Ponieważ ustawienie linijki wyżej (options.inJustDecodeBounds = true) pozwala pobrać dane z dekodowania bez tworzenia instancji. Czyli stawiamy "ramy", które pozwolą ustalić nam coś na temat zasobu zamiast tworzyć specjalnie obiekt z niego i wtedy pobierać dane. Dane zostaną ustawione w opcjach bez zbędnego żarcia pamięci. Dlatego tuż przed kolejnym wywołaniem decodeResource zmieniamy parametr options.inJustDecodeBounds na "false" i wtedy dopiero tworzymy obiekt.